libcaf_net/caf/net/http/header.cpp

```diff
@@ -0,0 +1,110 @@
+// This file is part of CAF, the C++ Actor Framework. See the file LICENSE in
+// the main distribution directory for license terms and copyright or visit
+// https://github.com/actor-framework/actor-framework/blob/master/LICENSE.
+
+#include "caf/net/http/header.hpp"
+
+#include "caf/logger.hpp"
+#include "caf/string_algorithms.hpp"
+
+namespace caf::net::http {
+
+namespace {
+
+constexpr std::string_view eol = "\r\n";
+
+// Iterates through lines separated by eol and applies f to them, until an empty
+// line is found (containing only eol). Returns the unprocessed input or error
+// if no eol is found or `f` returns false.
+template <class F>
+expected<std::string_view> process_lines(std::string_view input, F&& f) {
+  for (;;) {
+    size_t pos = input.find(eol);
+    // Stop when not finding the delimiter.
+    if (pos == std::string_view::npos)
+      return make_error(sec::logic_error, "EOL delimiter not found");
+    // Stop at the first empty line and return the remainder.
+    if (pos == 0) {
+      input.remove_prefix(eol.size());
+      return {input};
+    }
+    // Stop if the predicate returns false.
+    if (!f(input.substr(0, pos)))
+      return make_error(sec::logic_error, "Predicate function failed");
+    // Drop the consumed line from the input for the next iteration.
+    input.remove_prefix(pos + eol.size());
+  }
+}
+
+} // namespace
+
+header::~header() {
+  // nop
+}
+
+header::header(const header& other) {
+  if (!other.raw_.empty()) {
+    raw_.assign(other.raw_.begin(), other.raw_.end());
+    reassign_fields(other);
+  } else {
+    raw_.clear();
+    fields_.clear();
+  }
+}
+
+header& header::operator=(const header& other) {
+  if (!other.raw_.empty()) {
+    raw_.assign(other.raw_.begin(), other.raw_.end());
+    reassign_fields(other);
+  } else {
+    raw_.clear();
+    fields_.clear();
+  }
+  return *this;
+}
+
+void header::reassign_fields(const header& other) noexcept {
```

Comment:
Shouldn't be declared noexcept because it calls `fields_.resize` (which might throw).

```suggestion
void header::reassign_fields(const header& other) {
```